qiskit/algorithms/minimum_eigensolvers/vqe.py

```diff
@@ -0,0 +1,380 @@
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""The variational quantum eigensolver algorithm."""
+
+from __future__ import annotations
+
+import logging
+from time import time
+from collections.abc import Callable, Sequence
+
+import numpy as np
+
+from qiskit.algorithms.gradients import BaseEstimatorGradient
+from qiskit.circuit import QuantumCircuit
+from qiskit.circuit.library import RealAmplitudes
+from qiskit.opflow import PauliSumOp
+from qiskit.primitives import BaseEstimator
+from qiskit.quantum_info.operators.base_operator import BaseOperator
+from qiskit.utils import algorithm_globals
+
+from ..exceptions import AlgorithmError
+from ..list_or_dict import ListOrDict
+from ..optimizers import Optimizer, Minimizer, SLSQP
+from ..variational_algorithm import VariationalAlgorithm, VariationalResult
+from .minimum_eigensolver import MinimumEigensolver, MinimumEigensolverResult
+
+# from qiskit.algorithms.observables_evaluator import eval_observables
+
+logger = logging.getLogger(__name__)
+
+
+class VQE(VariationalAlgorithm, MinimumEigensolver):
+    r"""The variational quantum eigensolver (VQE) algorithm.
+
+    VQE is a quantum algorithm that uses a variational technique to find the minimum eigenvalue of
+    the Hamiltonian :math:`H` of a given system [1].
+
+    An instance of VQE requires defining two algorithmic sub-components: a trial state (a.k.a.
+    ansatz) which is a :class:`QuantumCircuit`, and one of the classical
+    :mod:`~qiskit.algorithms.optimizers`.
+
+    The ansatz is varied, via its set of parameters, by the optimizer, such that it works towards a
+    state, as determined by the parameters applied to the ansatz, that will result in the minimum
+    expectation value being measured of the input operator (Hamiltonian).
+
+    The optimizer can either be one of Qiskit's optimizers, such as
+    :class:`~qiskit.algorithms.optimizers.SPSA` or a callable with the following signature:
+
+    .. note::
+
+        The callable _must_ have the argument names ``fun, x0, jac, bounds`` as indicated
+        in the following code block.
+
+    .. code-block:: python
+
+        from qiskit.algorithms.optimizers import OptimizerResult
+
+        def my_minimizer(fun, x0, jac=None, bounds=None) -> OptimizerResult:
+            # Note that the callable *must* have these argument names!
+            # Args:
+            #     fun (callable): the function to minimize
+            #     x0 (np.ndarray): the initial point for the optimization
+            #     jac (callable, optional): the gradient of the objective function
+            #     bounds (list, optional): a list of tuples specifying the parameter bounds
```

Comment:
One more thought: I do not see the `bounds` being used anywhere in the code below.. Am I overlooking something?

Reply:
Bounds was fetched from an ansatz or some default if the ansatz did not not supply and passed to the optimizer/minimizer call. This aspect seems to not be present here as you noted - ie no code for bounds

Reply:
I assume that this needs to be re-added, correct?

Reply:
I've re-added this logic. One thing that seems a little strange to me in the original logic is that the bounds logic is handled after the initial point, which assumes (-2pi, 2pi) if not otherwise stated. So, if no bounds are set in the ansatz, we assume (-2pi, 2pi) to set the random initial point, but use [(None, None), ...] for the optimization. Should we instead use the bounds we assume in the `_validate_initial_point`.

Reply:
The `None`-default is actually what causes this issue: https://github.com/Qiskit/qiskit-nature/issues/818 and is the reason I noticed this behavior.
I do not have an ideal solution here, but seeing that in the past the random sample would be bound to `(-2pi, 2pi)` and assuming that this was correct/intended, I would use that same default for the optimization, hopefully also fixing that issue above..

What do you think?

Reply:
I think using `(-2pi, 2pi)` for the optimization makes sense. I'll go with that for now. Unless you know a reason this shouldn't be implemented, @woodsp-ibm?

Reply:
The random initial point was chosen within the bounds from the ansatz; if it was all None or any upper/lower value within the bound was None then the values were set to -2pi/2pi respectively to bound where the initial point was chosen. The bounds from the anstaz is passed on, whether None or not. The only thing it had was a sanity check on the length, if non None, before passing that on. We used to do a bit different in the early days where bounds was also checked against optimizer capabilities https://github.com/Qiskit/qiskit-aqua/blob/7a32ec864fbf48818a84be98af6cfbfb07d5dd9f/qiskit_acqua/vqe/vqe.py#L275 I am not sure exactly when things changed.
Arguably as I mentioned, given the behavior is often cyclic allowing the initial point to sit on the boundary can lead to things getting stuck #5713

Reply:
Having spoken more with @woodsp-ibm, it seems inappropriate to set an ansatz-independent bound, so I'll keep the logic the same as before. I suppose we'll have to address [Qiskit/qiskit-nature#818](https://github.com/Qiskit/qiskit-nature/issues/818) in another way.